engine/execution/ingestion/engine.go

```diff
@@ -555,7 +564,7 @@ func (e *Engine) executeBlock(ctx context.Context, executableBlock *entity.Execu
 
 	isExecutedBlockSealed := executableBlock.Block.Header.Height <= lastSealed.Height
 	broadcasted := false
-	if !isExecutedBlockSealed {
+	if !isExecutedBlockSealed && e.staker.AmIStaked() {
```

Comment:
Not sure how this works. Isn't determining whether I'm staked depending on the block ID? 
```suggestion
	if !isExecutedBlockSealed && e.staker.AmIStaked(blockID) {
```


If I'm no longer staked at the latest block, but I'm still executing blocks from the previous epoch which I'm still staked, then I will not broadcast receipts?

Reply:
We want to avoid checking stuff every block because epoch changes unfrequently. 
But also we don't know absolutely at the time of executing, because blocks are not sealed. 
This way we may broadcast a few blocks when we are not staked, but it's better than not broadcasting when we should be.
I think @jordanschalm could help us here, as entire epoch transitions is a bit tricky.

Reply:
are we able to cache the height where i'm going to stake so that it becomes cheaper to get the accurate result:
```
e.staker.AmIStaked(block.Height)
```

Reply:
Leo is right, we should always use the `blockID` as a reference point when checking the identity table state. 

> We want to avoid checking stuff every block because epoch changes unfrequently.

This is true as well. The identity table changes very infrequently so it is well-suited for caching. We should be able to read the identity table state from a cache for practically every query. The question is where this cache should live.

Essentially the `Staker` component introduced in this PR is a cache of a fairly small subset of the identity table: the current stake for the node's own identity. This works for this case, but the need for efficient access to the identity table is quite a bit broader than this case and I think it would be better suited by a single cache equivalent in scope to the API exposed by the protocol state. For example, we also will need to frequently query whether an arbitrary node which has sent us a message is staked with respect to the block referenced by the message, when deciding whether or not to process the message.

Therefore I suggest we use the protocol state `Identities` method directly for these queries. Even with epoch transitions and without slashing, it should already be sufficiently performant -- requests for the last few thousand blocks should always hit the cache. As we introduce more places where the identity table can change, we should apply the necessary caching / performance improvements to efficiently query identities to the protocol state directly.

Reply:
Thanks Jordan, that's very insightful!
I think it's still worth to keep a separate `Staker` component to hide the complexity of querying staking state from engine, but I will make sure it queries protocol state for a block

Reply:
Would this work? This is what Yahya used for the verification node. https://github.com/onflow/flow-go/blob/60d96cae587c604a61fcb77b8325d8a56133e1fe/state/protocol/state.go#L67-L75

Reply:
Yes, pretty much the same code, but wrapped in a separate component - it will be easier to mock.

Reply:
Another way to make it easier to mock would be to extract out the function as a field of the `Engine` that can be injected by the tests:
```go
type Engine struct {
  // ...
  amIStaked: func(blockID flow.Identifier) (bool, error)
}
```

or preferably create a helper for the testing code to do this:
```go
func (suite *TestSuite) mockMyIdentityAtBlock(blockID flow.Identifier, id *Identity) { ... }
```

Given that it only wraps one function call, I think this component is quite far from the threshold where adding a layer of abstraction or compartmentalization is beneficial.

Reply:
Is it possible that we don't mock? We just need to make sure in the tests that my own identity will have some stake, which shouldn't be hard to fix, no?

If so, we could simply reuse the [`IsNodeStakedAtBlockID`](https://github.com/onflow/flow-go/blob/60d96cae587c604a61fcb77b8325d8a56133e1fe/state/protocol/state.go#L67-L75)

Reply:
> Is it possible that we don't mock? We just need to make sure in the tests that my own identity will have some stake, which shouldn't be hard to fix, no?

It is always possible. But now we need to mock external condition irrelevant to majority of tests, specifying staking state indirectly.

Reply:
> Another way to make it easier to mock would be to extract out the function as a field of the Engine that can be injected by the tests:

This function would be stateful- it needs to access instance of `state` either by closing its value or having extra parameter.